File: 0204-count-primes/0204-count-primes.ts
function countPrimes(n: number): number {
    if (n <= 2) {
        return 0;
    }

    const prime: boolean[] = new Array(n).fill(true);
    prime[0] = false;
    prime[1] = false;

    let p = 2;
    while (p * p <= n) {
        if (prime[p]) {
            for (let i = p * p; i < n; i += p) {
                prime[i] = false;
            }
        }
        p++;
    }

    let count = 0;
    for (const isPrime of prime) {
        if (isPrime) {
            count++;
        }
    }

    return count;
}

File: 0204-count-primes/0204-count-primes.py
class Solution:
    def countPrimes(self, n: int) -> int:
        if n <= 2:
            return 0

        prime = [True] * n
        prime[0] = False
        prime[1] = False

        p = 2
        while p * p <= n:
            if prime[p]:
                for i in range(p * p, n, p):
                    prime[i] = False
            p += 1

        count = 0
        for is_prime in prime:
            if is_prime:
                count += 1

        return count
